src/Microsoft.AspNetCore.Mvc.Core/ApiExplorer/ApiConventionResult.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ApiExplorer
+{
+    /// <summary>
+    /// Metadata associated with an action method via API convention.
+    /// </summary>
+    public sealed class ApiConventionResult
+    {
+        public ApiConventionResult(IReadOnlyList<IApiResponseMetadataProvider> responseMetadataProviders)
+        {
+            ResponseMetadataProviders = responseMetadataProviders ??
+                throw new ArgumentNullException(nameof(responseMetadataProviders));
+        }
+
+        public IReadOnlyList<IApiResponseMetadataProvider> ResponseMetadataProviders { get; }
+
+        internal static ApiConventionResult GetApiConvention(MethodInfo method, ApiConventionTypeAttribute[] apiConventionAttributes)
+        {
+            foreach (var attribute in apiConventionAttributes)
+            {
+                var conventionMethod = GetConventionMethod(method, attribute.ConventionType);
+                if (conventionMethod != null)
+                {
+                    var metadataProviders = conventionMethod.GetCustomAttributes(inherit: false)
+                        .OfType<IApiResponseMetadataProvider>()
+                        .ToArray();
+
+                    return new ApiConventionResult(metadataProviders);
+                }
+            }
+
+            return null;
+        }
+
+        private static MethodInfo GetConventionMethod(MethodInfo method, Type conventionType)
+        {
+            foreach (var conventionMethod in conventionType.GetMethods())
+            {
+                if (IsMatch(method, conventionMethod))
+                {
+                    return conventionMethod;
+                }
+            }
+
+            return null;
+        }
+
+        internal static bool IsMatch(MethodInfo methodInfo, MethodInfo conventionMethod)
+        {
+            var methodNameMatchBehavior = GetNameMatchBehavior(conventionMethod);
+            if (!IsNameMatch(methodInfo.Name, conventionMethod.Name, methodNameMatchBehavior))
+            {
+                return false;
+            }
+
+            var methodParameters = methodInfo.GetParameters();
+            var conventionMethodParameters = conventionMethod.GetParameters();
+
+            for (var i = 0; i < conventionMethodParameters.Length; i++)
```

Comment:
Add a large comment here about what's allowed and what's disallowed. Also, in this method would probably aid in clarity and unit testability if you broke things down to 
```C#
if (!MethodMatches(...))
{
    return false;
}

if (!ParametersMatch(...))
{
    return false;
}

return true;
```